presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSource.java

```diff
@@ -197,38 +203,24 @@ public synchronized boolean appendToNext(PageBuilder pageBuilder, int outputChan
     @ThreadSafe
     private static class OuterPositionTracker
     {
-        private final List<Supplier<LookupSource>> lookupSourceSuppliers;
-
-        @GuardedBy("this")
-        private final boolean[][] visitedPositions;
-
-        @GuardedBy("this")
+        private final boolean[] visitedPositions;
         private boolean finished;
 
-        public OuterPositionTracker(List<Supplier<LookupSource>> lookupSourceSuppliers)
+        public OuterPositionTracker(int positionCount)
         {
-            this.lookupSourceSuppliers = lookupSourceSuppliers;
-            visitedPositions = new boolean[lookupSourceSuppliers.size()][];
-            for (int source = 0; source < lookupSourceSuppliers.size(); source++) {
-                try (LookupSource lookupSource = lookupSourceSuppliers.get(source).get()) {
-                    visitedPositions[source] = new boolean[lookupSource.getJoinPositionCount()];
-                }
-            }
+            visitedPositions = new boolean[positionCount];
         }
 
-        public synchronized void positionVisited(int partition, int position)
+        public synchronized void positionVisited(int position)
```

Comment:
I do not like exposing visitedPositions arrays from position trackers.
I know this is stop-gap but it seems it can be done simpler by changing
`positionVisited` to:
```java
public void positionVisited(int position) {
    verify(!finished);
    synchronised(visitedPositions[partition]) {
             visitedPositions[partition][position] = true;
    }
}
```

And `getOuterPositionIterator` to:
 
```java
public OuterPositionIterator getOuterPositionIterator() {
    finished = true;
    LookupSource[] lookupSources = lookupSourceSuppliers.stream()           
                     .map(Supplier::get)		
                     .toArray(LookupSource[]::new);
     for (int i = 0; i < visitedPartitions.size; ++i) {
        synchronized(visitedPartitions[i]) {} // just a memory barrier
     }
 		
      return new PartitionedLookupOuterPositionIterator(lookupSources, visitedPositions);
}
```

`finished` flag would probably need to be made volatile.